File: Data Structures & Algorithms/majority-element-ii/submission-2.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        candidates = defaultdict(int)
        for num in nums:
            candidates[num] += 1

            if len(candidates) <= 2:
                continue

            new_candidates = defaultdict(int)
            for (num, c) in candidates.items():
                if candidates[num] > 1:
                    new_candidates[num] = c - 1
            candidates = new_candidates

        res = []
        for num in candidates:
            if nums.count(num) > len(nums) // 3:
                res.append(num)

        return res

File: Data Structures & Algorithms/single-number-iii/submission-2.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int xors = 0;
        vector<int> res;
        for (int i = 0; i < nums.size(); i++) {
            xors = xors ^ nums[i];
        }
        int diff = xors & (-xors);
        int a = 0, b = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] & diff) {
                a = a ^ nums[i];
            }
            else {
                b = b ^ nums[i];
            }
        }
        return {a, b};
    }
};